cdap-unit-test/src/test/java/io/cdap/cdap/test/app/AppUsingGetServiceURL.java

```diff
@@ -107,6 +107,31 @@ public void ping(HttpServiceRequest request, HttpServiceResponder responder) thr
       }
     }
 
+    @GET
+    @Path("appWithServices")
+    public void appWithServices(HttpServiceRequest request, HttpServiceResponder responder) throws IOException {
+      URL serviceURL = getContext().getServiceURL("system", AppWithServices.APP_NAME, AppWithServices.SERVICE_NAME);
```

Comment:
I think it's better not to couple the two apps together like this. For the purposes of your test, you could add the same endpoint here and just deploy this app to two different namespaces.

The javadoc says it returns null if the service can't be discovered, but there are several layers to that. Does it return null if the app doesn't exist? When the app exists but the service isn't running yet? If the namespace doesn't exist?

In order to test these more naturally, I think it's better if the namespace comes from the request, either as part of the path or as a request header.

Reply:
I updated this code with your suggestion